_includes/parse-server/class-level-permissions.md

```diff
@@ -29,3 +29,255 @@ If you want to restrict access to a full class to only authenticated users, you
+
+```js
+{
+  classLevelPermissions:
+  {
+    "protectedFields": {
+      "*": ["views", "secret", "ownerEmail", "owner", "article",],
+      "authenticated": ["secret", "ownerEmail", "owner"]
+    }
+  }
+}
+```
+
+```js
+// example response for user without session token:
+{
+ "preview": "Lorem ipsum",
+}
+
+// example response for logged in user:
+{
+ "preview": "Lorem ipsum",
+ "article": "Lorem ipsum dolor sit amet",
+ "views": "42"
+}
+```
+
+It is essential to understand the basic principle how server determines which fields to protect when user belongs to multiple groups with different rules defined. First, server finds all scopes with `protectedFields` the user belongs to. Then resulting set is determined as an intersection of all applicable sets.
+
+In the above example, for logged in user:
+
+1. Both `*` and `authenticated` are applicable scopes.
+2. The result of intersection is `["secret", "ownerEmail", "owner"]` (the fields appear in both sets).
+3. Object will have all keys except `secret`, `ownerEmail` and `owner`.
+
+
+### `role:`
+
+To target users belonging to some role, use `role:role_name` as a key.
+
+Let's say we want to allow users with `admin` role to see all fields, while protecting some fields for all other users:
+
+```js
+{
+  classLevelPermissions:
+  {
+    "protectedFields": {
+      "*": ["ownerEmail", "secret"],
+      "role:admin": []
+    }
+  }
+}
+```
+
+In the above example we explicitly set empty array `[]` to state that no fields should be protected for users with `admin` role. So for `admin`, according to the principle discussed earlier:
+
+* Both `*` and `role:admin` scopes are applicable. 
+* The intersection result is `[]` (since there is no field that is present in both sets) thus all fields will appear in response.
+
+### Role hierarchy
+
+Sometimes your roles are related and form a hierarchy, e.g. when `tester` is related to `moderator` - `tester` inherits all privileges of `moderator`. Role hierarchy is also considered when protected fields are evaluated.
+
+```js
+let moderator = ... // Parse.Role
+let tester = ...    // Parse.Role
+
+moderator.getRelation('roles).add(tester);
+``` 
+
+Here is an example of a tricky setup that may lead to unexpected result, we'll explain why right after:
+
+```js
+{
+  classLevelPermissions:
+  {
+    "protectedFields": {
+      "role:moderator": ["secret"],
+      "role:tester": ["ownerEmail"]
+    }
+  }
+}
+```
+
+* When user with `role:moderator` fetches an object, `secret` is protected.
+* When user with `role:tester` fetches same object - all fields appear to be visible, even though `role:tester` has `ownerEmail` set as protected. This happens because of role hierarchy - when user has a role, he also implicitly gets all the inherited roles. Then server intersects sets for all roles (both `role:tester` and inherited `role:moderator` in this case) and intersection of `["secret"]` vs `["ownerEmail"]` results in `[]` (sets have no fields in common).
```

Comment:
```suggestion
* When a user with `role:tester` fetches the same object - all fields appear to be visible, even though `role:tester` has `ownerEmail` set as protected. This happens because of role hierarchy - when a user has a role, they also implicitly get all the inherited roles. Then server intersects sets for all roles (both `role:tester` and inherited `role:moderator` in this case) and intersection of `["secret"]` vs `["ownerEmail"]` results in `[]` (sets have no fields in common).
```